src/libstore/local-store.cc

```diff
@@ -1422,16 +1426,34 @@ Path LocalStore::addToStore(const string & name, const Path & _srcPath,
     Path srcPath(absPath(_srcPath));
     debug(format("adding ‘%1%’ to the store") % srcPath);
 
-    /* Read the whole path into memory. This is not a very scalable
-       method for very large paths, but `copyPath' is mainly used for
-       small files. */
-    StringSink sink;
-    if (recursive)
-        dumpPath(srcPath, sink, filter);
-    else
-        sink.s = readFile(srcPath);
+    Hash h = recursive ? hashPath(hashAlgo, srcPath, filter).first : hashFile(hashAlgo, srcPath);
+    Path dstPath = makeFixedOutputPath(recursive, hashAlgo, h, name);
+    addTempRoot(dstPath);
+    if (repair || !isValidPath(dstPath)) {
+
+        PathLocks outputLock(singleton<PathSet, Path>(dstPath));
+
+        if (repair || !isValidPath(dstPath)) {
+
+            if (pathExists(dstPath)) deletePath(dstPath);
+
+            copyPath(srcPath, dstPath, filter, recursive);
 
-    return addToStoreFromDump(sink.s, name, recursive, hashAlgo, repair);
+            canonicalisePathMetaData(dstPath, -1);
+            HashResult hash = hashPath(htSHA256, dstPath);
+            optimisePath(dstPath); // FIXME: combine with hashPath()
+
+            ValidPathInfo info;
+            info.path = dstPath;
+            info.hash = hash.first;
+            info.narSize = hash.second;
+            registerValidPath(info);
+        }
+
+        outputLock.setDeletion(true);
+    }
+
+    return dstPath;
```

Comment:
Only a minor thing:

How about reducing scope lengths by inverting that outer if-condition and returning early, so the reader needs to track less state in their head:

(also giving `repair` type `const bool` helps seeing that no side-effects will happen to it)

```c++
    if (!repair && isValidPath(dstPath)) {
        return dstPath;
    }

    PathLocks outputLock(singleton<PathSet, Path>(dstPath));

    if (!isValidPath(dstPath)) {

        if (pathExists(dstPath)) deletePath(dstPath);

        copyPath(srcPath, dstPath, filter, recursive);
        canonicalisePathMetaData(dstPath, -1);
        HashResult hash = hashPath(htSHA256, dstPath);
        optimisePath(dstPath); // FIXME: combine with hashPath()

        ValidPathInfo info;
        info.path = dstPath;
        info.hash = hash.first;
        info.narSize = hash.second;
        registerValidPath(info);
    }

    outputLock.setDeletion(true);

    return dstPath;
```